File: 732-my-calendar-iii/732-my-calendar-iii.go
import "github.com/emirpasic/gods/maps/treemap"


type MyCalendarThree struct {
    // Bookings map[int]int
    Bookings *treemap.Map
}


func Constructor() MyCalendarThree {
    // return MyCalendarThree{make(map[int]int)}
    return MyCalendarThree{treemap.NewWithIntComparator()}
}


func (this *MyCalendarThree) Book(start int, end int) int {
    count, ok := this.Bookings.Get(start)
    if !ok {
        count = 0
    }
    this.Bookings.Put(start, count.(int) + 1)

    count, ok = this.Bookings.Get(end)
    if !ok {
        count = 0
    }
    this.Bookings.Put(end, count.(int) - 1)
    
    max, sweepcount := 0, 0
    for _, count := range this.Bookings.Values() {
        sweepcount += count.(int)
        if sweepcount > max {
            max = sweepcount
        }
    }
    return max
}


/**
 * Your MyCalendarThree object will be instantiated and called as such:
 * obj := Constructor();
 * param_1 := obj.Book(start,end);
 */

File: 732-my-calendar-iii/732-my-calendar-iii.js

var MyCalendarThree = function() {
    this._treemap = {}
};

/** 
 * @param {number} start 
 * @param {number} end
 * @return {number}
 */
MyCalendarThree.prototype.book = function(start, end) {
    this._treemap[start] = (this._treemap[start] || 0) + 1
    this._treemap[end] = (this._treemap[end] || 0) - 1
    
    let sweepcount = 0
    let max = 0
    
    for (const count of Object.values(this._treemap)) {
        sweepcount += count
        if (sweepcount > max) {
            max = sweepcount
        }
    }
    return max
};

/** 
 * Your MyCalendarThree object will be instantiated and called as such:
 * var obj = new MyCalendarThree()
 * var param_1 = obj.book(start,end)
 */
